src/software/thunderscope/robot_diagnostics/handheld_device_manager.py

```diff
@@ -0,0 +1,393 @@
+        :return: the current status of the handheld device connection
+        """
+        return (
+            HandheldDeviceConnectionStatus.CONNECTED
+            if self.handheld_device is not None
+            else HandheldDeviceConnectionStatus.DISCONNECTED
+        )
+
+    def __send_proto_command_latest_primitive_controls(self) -> None:
+        self.proto_unix_io.send_proto(MotorControl, self.motor_control)
+        self.proto_unix_io.send_proto(PowerControl, self.power_control)
+
+        # reset power control
+        self.__initialize_default_power_control_values()
+
+    def refresh(self, mode: ControlMode):
+        """Refresh this class.
+        Spawns a new thread that runs the handheld device event
+        processing function if Control.Mode is Diagnostics,
+        otherwise, stops and joins the thread, if it is running
+        :param mode: The current user requested mode for controlling the robot
+        """
+        if mode == ControlMode.DIAGNOSTICS:
+            if self.__handheld_device_event_loop_thread.is_alive():
+                self.logger.debug("Terminating handheld device event handling process")
+                self.__shutdown_event_listener_thread()
+        elif mode == ControlMode.HANDHELD:
+            if not self.__handheld_device_event_loop_thread.is_alive():
+                self.logger.debug(
+                    "Setting up new handheld device event handling process"
+                )
+                self.__setup_new_event_listener_thread()
+                self.__start_event_listener_thread()
+
+            self.proto_unix_io.send_proto(MotorControl, self.motor_control)
+            self.proto_unix_io.send_proto(PowerControl, self.power_control)
+
+    def __clear_handheld_device(self) -> None:
+        """Clears handheld device & config field by setting to null,
+        and emits a disconnected notification signal.
+        """
+        self.handheld_device = None
+        self.handheld_device_config = None
+        self.handheld_device_disconnected_signal.emit(
+            HandheldDeviceConnectionStatus.DISCONNECTED
+        )
+
+    def __shutdown_event_listener_thread(self) -> None:
+        """Shut down the event processing loop by setting the stop event flag,
+        and then joins the handling thread, if it is alive.
+        """
+        # TODO (#3165): Use trace level logging here
+        # self.logger.debug("Shutdown down handheld device event loop process")
+        self.__stop_thread_signal_event.set()
+        if self.__handheld_device_event_loop_thread.is_alive():
+            self.__handheld_device_event_loop_thread.join()
+        self.__stop_thread_signal_event.clear()
+
+    def __start_event_listener_thread(self) -> None:
+        """Starts the thread that runs the event processing loop."""
+        self.__handheld_device_event_loop_thread.start()
+
+    def __setup_new_event_listener_thread(self):
```

Comment:
`__setup` and `__start` could be the same function tbh. We'll always call them together